drivers/siemens-sitop.c

```diff
@@ -0,0 +1,314 @@
+					current_ups_status.dc_input_low = 0;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "DC_LO", 5) == 0) {
+					current_ups_status.dc_input_low = 1;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "*****", 5) == 0) {
+					current_ups_status.on_battery = 0;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "*BAT*", 5) == 0) {
+					current_ups_status.on_battery = 1;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "BA>85", 5) == 0) {
+					current_ups_status.battery_above_85_percent = 1;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else if (strncmp(rx_buffer, "BA<85", 5) == 0) {
+					current_ups_status.battery_above_85_percent = 0;
+					rm_buffer_head(5 + NR_DELIMITER_CHARS);
+					new_data_received = 1;
+				} else {
+					/* nothing sensible found at the start of the rx_buffer. */
+					rm_buffer_head(1);
+				}
+			}
+		}
+	}
+	return new_data_received;
+}
+
+
+static int instcmd(const char *cmdname, const char *extra) {
+	/* See the remarks in upsdrv_shutdown() */
+	if (!strcasecmp(cmdname, "shutdown.return")) {
+		upslogx(LOG_NOTICE, "instcmd: sending command R");
+		ser_send_pace(upsfd, 200000, "\n\nR\n\n");
+		ser_send_pace(upsfd, 200000, "R\n\n");
+		return STAT_INSTCMD_HANDLED;
+	}
+	if (!strcasecmp(cmdname, "shutdown.stayoff")) {
+		upslogx(LOG_NOTICE, "instcmd: sending command S");
+		ser_send_pace(upsfd, 200000, "\n\nS\n\n");
+		ser_send_pace(upsfd, 200000, "S\n\n");
+		return STAT_INSTCMD_HANDLED;
+	}
+
+	upslogx(LOG_NOTICE, "instcmd: unknown command [%s]", cmdname);
+	return STAT_INSTCMD_UNKNOWN;
+}
+
+
+void upsdrv_initinfo(void) {
+	int max_attempts = 5;
+	int found = 0;
+	while (!found && max_attempts > 0) {
+		if (check_for_new_data())
+			found = 1;
+		else
+			sleep(1); /* Sleep a while, then try again */
+		max_attempts--;
+	}
+	if (!found)
+		fatalx(EXIT_FAILURE, "No data received from the UPS");
+	
+	dstate_setinfo("ups.mfr", "Siemens");
+	dstate_setinfo("ups.model", "SITOP UPS500 series");
+	/* note: for a transition period, these data are redundant! */
+	dstate_setinfo("device.mfr", "Siemens");
+	dstate_setinfo("device.model", "SITOP UPS500 series");
+	
+	/* supported commands: */
+	dstate_addcmd("shutdown.stayoff");
+	dstate_addcmd("shutdown.return");
+
+	upsh.instcmd = instcmd;
+}
+
+
+void upsdrv_updateinfo(void) {
+	if (check_for_new_data())
+		nr_polls_without_data = 0;
+	else {
+		nr_polls_without_data++;
+		if (nr_polls_without_data < 0)
+			nr_polls_without_data = INT_MAX;
+	}
+	
+	if (nr_polls_without_data > max_polls_without_data) {
+		/* data is stale */
+		dstate_datastale();
+		return;
+	}
+	
+	/* This is all we know about the charge level... */
+	dstate_setinfo("battery.charge.indication",
```

Comment:
This is a blocker. If you add a new var you are supposed to:
- document it in nut-names.txt / cmdvartab
- get it approved.

Reply:
I have considered to use the existing variable battery.charge, but people (and code?) would probably assume that it contains an integer (percentage), not a string ("<85%" or ">85%").
As far as I could see, the list of existing variables does not contain a variable that suits this purpose, or does it?